### PR TITLE
General Proof calculation algorithm

### PR DESCRIPTION
**Pull request by gsora**
Instead of relying on two different structs to physically split verified and unverified DID Documents, employ a logical separation and assert unverified data when calling `VerifyProof()`:

```go
type DidDocument struct {
        // ...
	// Proof is **NOT** optional, we need it to have omitempty to make the signature procedure more straightforward,
	// i.e. DidDocument.Validate() will check if proof is empty, and throw an error if true.
	Proof *Proof `json:"proof,omitempty"`
        // ...
}

// DidDocumentUnsigned is an intermediate type used to check for proof correctness
// It is identical to a DidDocument, it's kept for logical compartimentization.
type DidDocumentUnsigned DidDocument

func (didDocument DidDocument) VerifyProof() error {
	u := DidDocumentUnsigned(didDocument)

	// Explicitly zero out the Proof field.
	//
	// Here we leverage the fact that encoding/json do not encode nil pointers,
	// effectively giving us DidDocument-(Proof field).
	u.Proof = nil

        // ...
}
```

This is a quite radical change, but I added tests to cover message processing corner cases, as well as updates to CLI commands.

The motivation behind this change is the following: in the future we might want to add/remove fields from DidDocument, having an almost-identical copy minus the Proof field might lead to one struct getting updated while the other one isn't. 

## Checklist
- [x] Targeted PR against correct branch
- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For Admin Use:
- [x] Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- [x] Reviewers Assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]"

### CODE REVIEW

File: x/id/types/did_document_test.go [hunk 1]
```diff
@@ -259,3 +259,382 @@ func getBaseDocumentWithServices(services []Service) DidDocument {
...
+			Controller:         testOwnerAddress.String(),
+			SignatureValue:     "uv9ZM4XusZl2q6Ei2O7aZW32pzwfg6ZQpBsQPb8cxzlFXWEyZLxem29fQBB4Py3W5gaXFEyPGruMXNsNDnr4sQ==",
+			VerificationMethod: "did:com:pub1addwnpepqwzc44ggn40xpwkfhcje9y7wdz6sunuv2uydxmqjrvcwff6npp2exy5dn6c",
+		},
+		PubKeys: PubKeys{
+			PubKey{
+				ID:         fmt.Sprintf("%s#keys-1", testOwnerAddress),
+				Type:       "RsaVerificationKey2018",
+				Controller: testOwnerAddress,
+				PublicKeyPem: `-----BEGIN PUBLIC KEY-----
```
Comment (alessio) [inline x/id/types/did_document_test.go:284]: Since this content is reused all over the place, maybe we could save it in a file somehwere under a `testdata` directory?

Comment (gsora) [reply x/id/types/did_document_test.go:284]: Agreed, I'll take care of it in the cleanup branch.
